### PR TITLE
fix: copy `ParamInfo` instances from signatures before processing them

### PR DESCRIPTION
**Pull request by EQUENOS**
## Summary

Currently it's not recommended to reference the same `ParamInfo` instance in multiple signatures, as it results in weird behavior of `LocalisationStore`. However, this is not documented and it's not obvious for an average user. Sometimes creating a single `ParamInfo` instance for multiple slash commands is very tempting, e.g. a choice between `On` and `Off` with localized labels. This PR allows to safely do that.

This example reproduces the bug I'm talking about:
```python
import disnake
from disnake.ext import commands

bot = commands.InteractionBot()
MY_PARAM: str = commands.Param(...)  # imagine a ton of arguments here

@bot.slash_command()
async def cmd_a(inter: disnake.GuildCommandInteraction, arg: str = MY_PARAM) -> None:
    """This is command A {{CMD_A}}

    Parameters
    ----------
    arg: Enter the argument {{CMD_A_ARG}}
    """
    ...

@bot.slash_command()
async def cmd_b(inter: disnake.GuildCommandInteraction, arg: str = MY_PARAM) -> None:
    """This is command B {{CMD_B}}

    Parameters
    ----------
    arg: Enter the argument {{CMD_B_ARG}}
    """
    ...
```

## Checklist

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by onerandomusername — CHANGES_REQUESTED** [review]
untested

File: disnake/ext/commands/params.py [hunk 1]
```diff
@@ -530,6 +530,30 @@ def __init__(
...
+        new.choices = self.choices.copy()
+        new.type = self.type
+        new.channel_types = self.channel_types.copy()
+        new.max_value = self.max_value
+        new.min_value = self.min_value
+        new.min_length = self.min_length
+        new.max_length = self.max_length
+        new.large = self.large
+
+        return new
```
Comment (onerandomusername) [inline disnake/ext/commands/params.py:555]:
```suggestion
        ins = cls.__new__(cls)

        ins.name = self.name
        ins.name_localizations = self.name_localizations._copy()
        ins.description = self.description
        ins.description_localizations = self.description_localizations._copy()
        ins.default = self.default
        ins.param_name = self.param_name
        ins.converter = self.converter
        ins.convert_default = self.convert_default
        ins.autocomplete = self.autocomplete
        ins.choices = self.choices.copy()
        ins.type = self.type
        ins.channel_types = self.channel_types.copy()
        ins.max_value = self.max_value
        ins.min_value = self.min_value
        ins.min_length = self.min_length
        ins.max_length = self.max_length
        ins.large = self.large

        return ins
```

File: disnake/ext/commands/params.py [hunk 2]
```diff
@@ -530,6 +530,30 @@ def __init__(
         self.max_length = max_length
         self.large = large
 
+    def copy(self) -> ParamInfo:
+        cls = self.__class__
```
Comment (onerandomusername) [inline disnake/ext/commands/params.py:534]:
```suggestion
        # n. b. this method needs to be manually updated with a new attribute is added.
        cls = self.__class__
```

**Review by onerandomusername — APPROVED** [review]
tested, lgtm

**Review by shiftinv — CHANGES_REQUESTED** [review]

File: disnake/ext/commands/params.py [hunk 3]
```diff
@@ -559,7 +584,8 @@ def from_param(
         parsed_docstring = parsed_docstring or {}
 
         if isinstance(param.default, cls):
-            self = param.default
+            # we copy this ParamInfo instance because it can be used in multiple signatures
+            self = param.default.copy()
```
[the remark this answers is not in the document]
Comment (EQUENOS) [reply disnake/ext/commands/params.py:588]: My bad, for some reason I thought that param parsing only happens once

File: disnake/i18n.py [hunk 4]
```diff
@@ -225,6 +225,13 @@ def _link(self, store: LocalizationProtocol) -> None:
         if self._key is not None:
             self._data = store.get(self._key)
 
+    def _copy(self) -> LocalizationValue:
+        cls = self.__class__
+        ins = cls.__new__(cls)
+        ins._key = self._key
+        ins._data = None if self._data is None else self._data.copy()
```
[the remark this answers is not in the document]
Comment (EQUENOS) [reply disnake/i18n.py]: Fixed in https://github.com/DisnakeDev/disnake/pull/946/commits/0ce581cc491172e35e35d374a3518fb9ea10d62e

File: disnake/ext/commands/params.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EQUENOS) [reply disnake/ext/commands/params.py:588]: Fixed in https://github.com/DisnakeDev/disnake/pull/946/commits/0ce581cc491172e35e35d374a3518fb9ea10d62e

File: disnake/ext/commands/params.py [hunk 5]
```diff
@@ -938,18 +968,18 @@ def collect_params(
             paraminfo = ParamInfo.from_param(parameter, {}, doc)
             paraminfos.append(paraminfo)
 
-    return (
+    function.__collect_params_cache__ = (
```
[the remark this answers is not in the document]
Comment (EQUENOS) [reply disnake/ext/commands/params.py]: https://github.com/DisnakeDev/disnake/pull/946/commits/fc46627a4687d3adb05bc02010937b91b05c7584

File: disnake/i18n.py [hunk 6]
```diff
@@ -225,6 +225,13 @@ def _link(self, store: LocalizationProtocol) -> None:
         if self._key is not None:
             self._data = store.get(self._key)
 
+    def _copy(self) -> LocalizationValue:
+        cls = self.__class__
+        ins = cls.__new__(cls)
+        ins._key = self._key
+        ins._data = self._data.copy() if isinstance(self._data, dict) else self._data
```
Comment (shiftinv) [inline disnake/i18n.py:232]: Hmm, I don't think we should copy this, as the reference returned by `LocalizationProtocol.get` is retained everywhere else. A shallow copy is fine here

The fact that modifying localization data works after everything was initialized is undocumented, but could still be useful to some.

Comment (EQUENOS) [reply disnake/i18n.py]: Changed this to shallow copy: https://github.com/DisnakeDev/disnake/pull/946/commits/2abd525839c05f7928150fab548a9ef5be8d15f8

**Review by shiftinv — APPROVED** [review]
lgtm, thanks!

**Pull request merged**